### ISSUE TITLE
no information about the cell border is given in an Excel sheet, if the cell border is different from the adjacent cells.

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#3044]
**Reported by sumandogra on 2013-03-04 11:34**
Excel 2010

NVDA does not announce the information regarding the border of a cell in any way.  In fact on pressing NVDA +F, it incorrectly announces the text as underlined in case the border of a cell is different from the adjacent cells.
Steps to reproduce:
1.  Go to a cell that has a different cell border than the default one.
2.  No information is given by NVDA about its different cell border.
3.  Press NVDA+F.
4.  NVDA announces it as underline.  This is an incorrect announcement and can lead to confusion at times to the users.

**Comment by nishimotz** [comment]
I would like to contribute regarding Excel cell border announcement, because it is frequently used in Japan.
Suggestions are welcomed on how cell borders should be announced in English.

some resources:
XlBordersIndex Enumeration
https://msdn.microsoft.com/en-us/library/office/ff835915.aspx
XlLineStyle Enumeration
https://msdn.microsoft.com/en-us/library/office/ff821622.aspx

**Comment by nishimotz** [comment]
I have worked regarding Excel cell borders at my repository:

https://github.com/nishimotz/nvda/tree/excel

Before proceeding to making pull request, I would like to hear comments.
So far, the descriptions in English are as follows.

```
black continuous bottom edge
B1
black continuous surrounding border
B2
B3
black continuous top edge
B4
black continuous up-right and down-right diagonal lines
B5
bright red continuous bottom edge
B6
bright red continuous surrounding border
B7
bright red continuous top edge, aqua-blue continuous bottom edge
B8
aqua-blue continuous surrounding border
B9 through C10
aqua-blue continuous top edge, black double bottom edge
B11
black double top edge
B12
black double bottom edge
B13
black double top edge, black continuous bottom edge
B14
black continuous top and bottom edges
B15
black continuous top edge
B16
black dash followed by two dots left and right edges
B17
no border lines
B18
black continuous left and right edges
B19
no border lines
B20
```

**Comment by nishimotz** [comment]
The example document (borders.xlsx) mentioned above is:
https://docs.com/nishimotz/8981/borders

**Comment by jcsteh** [comment]
@QChristensen, thoughts?

**Comment by Qchristensen** [comment]
I agree with the comments, having cell border information read much as @nishimotz suggests, for both NVDA+f and also having a document formatting option to read changed border information for announce formatting changes after the cursor, would be consistent with other formatting information read.

The first thing I thought when reading this ticket was, how is this done with text or table borders in Word, as it should present as consistent to the user.  It isn't currently available in Word, so it may be worth considering whether this can be implemented in both at the same time.

Reporting borders in Word is already filed specifically in issue #2010.

**Comment by Qchristensen** [comment]
With your descriptions @nishimotz, did you get those from Excel anywhere?  Looking in the borders page in the format cells dialog, (control+1, then b), the different types of borders are identified there.  The descriptions are slightly different to yours (for instance B16 would be "Dash, dot, dot" rather than "dash followed by two dots".  Also the continuous lines would be "thick", "medium" or "thin" rather than "continuous").

I just wanted to check in case you are getting your descriptions from somewhere else in Excel perhaps?

Looking for where else they might be reference, I noticed that in the border line style from the ribbon (press alt+h then b then y), you get a list of the styles which looks the same as in the format cells dialog, however NVDA reads each these simply as "list item" (Narrator simply identifies the place in the list).

**Comment by nishimotz** [comment]
The descriptions of my tentative work come from MSDN as follows:
https://msdn.microsoft.com/en-us/library/office/ff821622.aspx
Actually it should be consistent with format cells dialog of Excel itself.

**Comment by Qchristensen** [comment]
Ah thank you.  I wondered whether there was more than one place that
specified how the border style was described...




-- 
22 Point

Web: http://www.22point.com.au
Facebook: https://www.facebook.com/22Point
Twitter: https://twitter.com/22PointApps

Check out our first app, RapiTap! - Tap targets fast & avoid decoys:
Adrenaline pumping, Challenging, Accessible!
https://play.google.com/store/apps/details?id=au.com.twentytwopoint.rapitap
Free trial:
https://play.google.com/store/apps/details?id=au.com.twentytwopoint.rapitapfree

**Comment by nishimotz** [comment]
XlBorderWeight Enumeration should be handled as well:
https://msdn.microsoft.com/en-us/library/office/ff197515.aspx

**Comment by Qchristensen** [comment]
Yes, the border style information in the format cells dialog gives some
information about each of those - it doesn't say continuous for instance,
but it does give the thickness as you just posted.


-- 
22 Point

Web: http://www.22point.com.au
Facebook: https://www.facebook.com/22Point
Twitter: https://twitter.com/22PointApps

Check out our first app, RapiTap! - Tap targets fast & avoid decoys:
Adrenaline pumping, Challenging, Accessible!
https://play.google.com/store/apps/details?id=au.com.twentytwopoint.rapitap
Free trial:
https://play.google.com/store/apps/details?id=au.com.twentytwopoint.rapitapfree

**Comment by nishimotz** [comment]
* Should I work regarding braille output as well?
* Should I add documentFormatting.reportBorderStyle config parameter and setting dialog option as well?
* So far, NVDAObjects/window/excel.py has inconsistency regarding line ending characters, so I am hesitating to work with. May I clean up the file regarding line ending before working for the issue?

updated my branch which describes as follows:

Sheet2 of https://docs.com/nishimotz/8981/borders

```
no border lines
None  B2

black hair surrounding border
Hair  B4

black dotted surrounding border
Dotted  B6

black dash dot dot surrounding border
DashDotDot  B8

black dash dot surrounding border
DashDot  B10

black dashed surrounding border
Dashed  B12

black thin surrounding border
Thin  B14

black medium dash dot dot surrounding border
MediumDashDotDot  D2

black slanted dash dot surrounding border
SlantedDashDot  D4

black medium dash dot surrounding border
MediumDashDot  D6

black medium dashed surrounding border
MediumDashed  D8

black medium surrounding border
Medium  D10

black thick surrounding border
Thick  D12

black double surrounding border
Double  D14
```

**Comment by nishimotz** [comment]
Now I am feeling cell border colors are sometimes too verbose.

Regarding setting dialog options, I first thought documentFormatting.reportColor should be respected regarding cell border description.

However, reportColor is now moved into the Font group of format settings, so I need another flag regarding border colors, if we need less verbose mode for cell borders.

For example, Table group should have 'border style' and 'border color' check-boxes, and the border color option should be disabled when 'border style' is unchecked.
Or should we use a combo-box for this?

**Comment by leonardder** [comment]
Honestly, I belief the defaults for this are too verbose. I think both reporting of styles and colors should be off in their initial state.
